site/content/blog/2020-09-07-improving-performance-through-load-testing.md

```diff
@@ -0,0 +1,83 @@
+
+## What is load testing?
+
+Load testing is the process of applying a certain amount of load to a target system in order to understand how it behaves.
+The more realistic the generated load is, the more accurate and insightful the final results will be.
+
+## Why does it matter?
+
+Load testing can be used to improve both the performance and the overall quality of your application by helping you to:
+
+- Identify performance bottlenecks in your system.
+- Determine the overall capacity of your application.
+- Prevent performance regressions between releases.
+- Discover insidious bugs that passed functional testing.
+
+As an example, these are some of non-performance related, but rather significant issues we've been able to detect while load testing:
+
+- Data races
+- Application crashes
+- Deadlocks
+- Memory leaks
+
+As you can see, running load tests can guard your software against problems of different nature. By simulating users you get a chance to realistically test a system before it hits production and potentially avoid several issues down the line.
+
+## How are we doing it?
+
+There are plenty of extremely valid load testing frameworks available; some of the best are free and open source (JMeter, Gatling, k6 to name a few). In our case, after a thorough examination, we decided to develop a custom, in-house tool. We accepted trading implementation time and complexity for improved simulation and finer control over the entire process. We've recently released [version 1.0](https://github.com/mattermost/mattermost-load-test-ng/releases/tag/v1.0.0) and we are quite happy with the choice we made. While the tool itself is not designed for general purpose use, it can still serve as a good example of how we tackled the problem of designing and developing a custom load testing engine.
+
+### Simulating the load
+
+As the definition says, the core of the process is load generation. However, in order for results to be meaningful, we need the software to mimic actual user behaviour. We achieve this by leveraging data coming from real users through the use of:
+
+- Server-side metrics (e.g. frequency of a given API call)
+- User telemetry data (e.g. frequency of a user event/action)
+
+Fortunately, here at Mattermost we have a convenient advantage: we use our own application in our day to day work. This means we are able to continuously collect a considerable amount of realistic metrics from our community servers. We can then use such data to model user behaviour. To do this we create a table of user's actions with their respective frequencies. This action-frequency map will feed the simulation process: a simulated user will randomly pick and perform an action based on its frequency so that most common actions will be picked more often.
```

Comment:
```suggestion
Fortunately, here at Mattermost we have a convenient advantage: we use our own application in our day-to-day work. This means we're able to continuously collect a considerable amount of realistic metrics from our community servers. We can then use such data to model user behavior. To do this we create a table of a user's actions with their respective frequencies. This action-frequency map will feed the simulation process: a simulated user will randomly pick and perform an action based on its frequency so that the most common actions will be picked more often.
```